quint: init at 0.25.1

bugarela opened the pull request:
Hello :octocat: 

This adds a package for Quint, a fairly new (3 years old) formal specification language based on TLA+. Quint is not mainstream, but it is actively maintained, has currently [941 GitHub stars](https://github.com/informalsystems/quint/stargazers) and a decent amount of downloads on [NPM](https://www.npmjs.com/package/@informalsystems/quint) and [VSCode](https://marketplace.visualstudio.com/items?itemName=informal.quint-vscode). I work on Quint as my full time job and will be a maintainer for this nix package.

Here's the official website: https://quint-lang.org/

Quint automatically downloads two dependencies (Apalache, the model checker in Scala (JVM) and its alternative evaluator in Rust (binary) - both released under Apache 2.0) if they are not found in the expected folders (under `$QUINT_HOME`). So what I'm doing in this package is building/downloading these dependencies in a nix-compatible way, and then Quint will not perform any automatic downloads.

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)

The command finishes successfully, but I get this report:
```
--------- Report for 'x86_64-linux' ---------
1 package built:
quint

error: build log of '/nix/store/1jy8bcpgmxsbgdhcmrqgn1mzl84h1cyz-quint-0.25.1.drv^*' is not available
```
I'm not sure how to proceed.

- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

bugarela committed:
quint: init at 0.25.1

Co-authored-by: emaryn <197520219+emaryn@users.noreply.github.com>

bugarela committed:
maintainers: add bugarela

zmitchell commented:
Commits probably need to be reversed so that you're a maintainer before adding the package.

bugarela commented on pkgs/by-name/qu/quint/package.nix:
```diff
@@ -0,0 +1,108 @@
+  version = quintVersion;
+
+  nativeBuildInputs = [ makeWrapper ];
+  buildInputs = [ jre ];
```
I was pretty sure we needed it in order to run Apalache, which is a `jar` and needs the JVM. I just tried removing it and it still works, but I don't understand why. I can be testing it wrong tho.

bugarela commented on pkgs/by-name/qu/quint/package.nix:
```diff
@@ -0,0 +1,108 @@
+    dontNpmPrune = true;
+    installPhase = ''
+      mkdir -p $out/share/quint
+      cp -r node_modules $out/share/quint/
```
We do, otherwise the dependencies can't be found. Those can be moves (`mv`) instead of copies though, I'll make that change.

bugarela commented on pkgs/by-name/qu/quint/package.nix:
```diff
@@ -0,0 +1,108 @@
+
+    useFetchCargoVendor = true;
+    cargoHash = "sha256-beWqUDaWWCbGL+V1LNtf35wZrIqWCCbFLYo5HCZF7FI=";
+  };
```
I've added the full meta for each of them. Let me know if you want me to keep the license only.

bugarela commented:
Thank you for the careful review @emaryn! I believe I addressed all of your comments. I can squash the package-related (not the `maintainers:` one) commits into one after we're done iterating - don't want to cause confusion with force-pushes at this point.

bugarela commented on pkgs/by-name/qu/quint/package.nix:
```diff
@@ -0,0 +1,136 @@
+  };
+
+  # Build the Quint CLI from source
+  quintCLI = buildNpmPackage {
```
It can, partially. Certain commands (like `quint run --backend=rust` and `quint verify`) will download dependencies for you automatically, which can break nix things. So, for nix, the only way to install this should be with the full package here, in my understanding of things.

bugarela commented on pkgs/by-name/qu/quint/package.nix:
```diff
@@ -0,0 +1,136 @@
+    name = "quint";
+
+    nativeBuildInputs = [
+      pkgs.nodejs_20
```
it doesn't work without pkgs, is there something else I should use? Or do you mean that I should use a pinned version of nixpkgs?

```
error:
       … while evaluating a branch condition
         at /home/gabriela/projects/nixpkgs/lib/customisation.nix:305:5:
          304|     in
          305|     if missingArgs == { } then
             |     ^
          306|       makeOverridable f allArgs

       … while calling the 'removeAttrs' builtin
         at /home/gabriela/projects/nixpkgs/lib/attrsets.nix:657:28:
          656|   */
          657|   filterAttrs = pred: set: removeAttrs set (filter (name: !pred name set.${name}) (attrNames set));
             |                            ^
          658|

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: undefined variable 'nodejs_20'
       at /home/gabriela/projects/nixpkgs/pkgs/by-name/qu/quint/package.nix:31:7:
           30|     nativeBuildInputs = [
           31|       nodejs_20
             |       ^
           32|     ];
```

bugarela commented on pkgs/by-name/qu/quint/package.nix:
```diff
@@ -0,0 +1,136 @@
+  jre,
+  fetchzip,
+  buildNpmPackage,
+  pkgs,
```
Thanks, that makes more sense :sweat_smile:

bugarela commented on pkgs/by-name/qu/quint/package.nix:
```diff
@@ -0,0 +1,136 @@
+}:
+
+let
+  quintVersion = "0.25.1";
+  apalacheVersion = "0.47.2";
+  evaluatorVersion = "0.2.0";
```
Sorry, I don't understand this one. I need them in the `let` statement as they are all used in the let definitions

bugarela commented on pkgs/by-name/qu/quint/package.nix:
```diff
@@ -0,0 +1,136 @@
+
+  # Build the Quint CLI from source
+  quintCLI = buildNpmPackage {
+    name = "quint";
```
Ok, I now have
```nix
name = "quint";
pname = "quint-cli";
```

Is that good? Again, I can't remove `name` as I get an error about the missing attribute.

bugarela commented:
> Will setting $QUINT_HOME have any side effects?

Not any undesirable ones. I've added this env var to Quint only for the nix package to be able to set it

bugarela commented on pkgs/by-name/qu/quint/package.nix:
```diff
@@ -0,0 +1,119 @@
+      --set QUINT_HOME "$out/.quint" \
+      --prefix PATH : ${lib.makeBinPath [ jre ]}
+
+    install -Dm755 ${quint-evaluator}/bin/quint_evaluator -t $out/.quint/rust-evaluator-v${evaluatorVersion}/
+
+    mkdir -p $out/.quint/apalache-dist-${apalacheVersion}
+    cp -r ${apalacheDist} $out/.quint/apalache-dist-${apalacheVersion}/apalache
+    chmod +x $out/.quint/apalache-dist-${apalacheVersion}/apalache/bin/apalache-mc
```
I can't because I need to `chmod +x` the bin files, so I need a copy of the ones from nix store (instead of a link)

pbsds commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 414584`
Commit: `0bc7143c2c7c24e8b81b4744c40ae68ad3f2c247`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>quint</li>
  </ul>
</details>
